pkg/trace/local_client.go

```diff
@@ -0,0 +1,145 @@
+package trace
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"github.com/cometbft/cometbft/config"
+	"github.com/cometbft/cometbft/libs/log"
+)
+
+// Event wraps some trace data with metadata that dictates the table and things
+// like the chainID and nodeID.
+type Event struct {
+	ChainID   string      `json:"chain_id"`
+	NodeID    string      `json:"node_id"`
+	Table     string      `json:"table"`
+	Timestamp time.Time   `json:"timestamp"`
+	Msg       interface{} `json:"msg"`
+}
+
+// NewEvent creates a new Event with the given chainID, nodeID, table, and msg.
+// It adds the current time as the timestamp.
+func NewEvent(chainID, nodeID, table string, msg interface{}) Event {
+	return Event{
+		ChainID:   chainID,
+		NodeID:    nodeID,
+		Table:     table,
+		Msg:       msg,
+		Timestamp: time.Now(),
+	}
+}
+
+// LocalClient saves all of the events passed to the retuen channel to files
+// based on their "type" (a string field in the event). Each type gets its own
+// file. The internals are purposefully not *explicity* thread safe to avoid the
+// overhead of locking with each event save. Only pass events to the returned
+// channel. Call CloseAll to close all open files.
+type LocalClient struct {
+	chainID, nodeID string
+	logger          log.Logger
+
+	mu sync.RWMutex // Protects access to the fileMap
```

Comment:
nit: generally the mutex should be directly over the fields it protects. If it doesn't protect the `cannal`, then this should be above the mutex

Reply:
Do we even need this mutex if all files are created ahead of time

Reply:
There's also a package called `autofile` which was used by the WAL and has a similar mechanism of creating new files everytime a certain file reaches a size (IIRC)

Reply:
fixed with new scheme https://github.com/celestiaorg/celestia-core/pull/1279/commits/b77ed97bbb8f8e3cb499696fbec4ce454ba00b19